libflux/src/semantic/nodes.rs

```diff
@@ -520,7 +520,40 @@ pub struct FunctionExpr {
 
 impl FunctionExpr {
     fn infer(&mut self, env: Environment, f: &mut Fresher) -> Result {
-        unimplemented!();
+        let mut env = env;
+        let mut cons: Vec<Constraint> = Vec::new();
+        let mut pipe = None;
+        let mut req = HashMap::new();
+        let mut opt = HashMap::new();
+        for fp in &mut self.params {
+            let (m, v) = match fp.default {
+                Some(ref mut e) => {
+                    let (nenv, ncons) = e.infer(env, f)?;
+                    cons.append(&mut Vec::from(ncons));
+                    env = nenv;
+                    (&mut opt, e.type_of().clone())
+                }
+                None => (&mut req, MonoType::Var(f.fresh())),
+            };
+
+            let k = fp.key.name.clone();
+            if fp.is_pipe {
+                pipe = Some(types::Property {
+                    k: k.clone(),
+                    v: v.clone(),
+                });
+            }
+            m.insert(k, v);
+        }
+
+        let (env, bcons) = self.body.infer(env, f)?;
```

Comment:
The environment you pass to the `infer` function here must contain the function parameters.